app/code/Magento/Catalog/Model/Product/Link/SaveHandler.php

```diff
@@ -55,17 +64,71 @@ public function execute($entityType, $entity)
     {
         $link = $entity->getData($this->metadataPool->getMetadata($entityType)->getLinkField());
         if ($this->linkResource->hasProductLinks($link)) {
-            /** @var \Magento\Catalog\Api\Data\ProductInterface $entity*/
+            /** @var \Magento\Catalog\Api\Data\ProductInterface $entity */
             foreach ($this->productLinkRepository->getList($entity) as $link) {
                 $this->productLinkRepository->delete($link);
             }
         }
         $productLinks = $entity->getProductLinks();
+
+        // Build links per type
+        $linksByType = [];
+        foreach ($productLinks as $link) {
+            $linksByType[$link->getLinkType()][] = $link;
+        }
+
+        // Do check
+        $hasPositionLinkType = $this->isPositionSet($linksByType);
```

Comment:
It's confusing that a variable name starting with `has` contains an array. The name suggests that it contains a boolean value.

Reply:
it contains an array of boolean values per link type, I had the case that the positions were set for the products in "Related" but not set for products under "Cross sell" and so on, 

the same that under on Link type got the case of some products had position attribute and others don't, therefor I had to check if at least one product doesn't have a position under a product Link type all positions under this type will be ignored and user array index instead. 

regarding the question weather we need that amount of iterations, 
This came from too many use cases and variants of sending data, as I said about some link types aren't used at all others have positions but not for all products, another type with full defined products positions.